pgjdbc/src/test/java/org/postgresql/test/jdbc2/StatementTest.java

```diff
@@ -411,6 +421,99 @@ public void testWarningsAreCleared() throws SQLException {
     stmt.close();
   }
 
+  @Test
+  public void testWarningsAreAvailableAsap()
+      throws SQLException, InterruptedException, ExecutionException {
+    con.createStatement()
+            .execute("CREATE OR REPLACE FUNCTION notify_then_sleep() RETURNS VOID AS "
+                + "$BODY$ "
+                + "BEGIN "
+                + "RAISE NOTICE 'Test 1'; "
+                + "RAISE NOTICE 'Test 2'; "
+                + "EXECUTE pg_sleep(2); "
+                + "END "
+                + "$BODY$ "
+                + "LANGUAGE plpgsql;");
+    con.createStatement().execute("SET SESSION client_min_messages = 'NOTICE'");
+    final PreparedStatement preparedStatement = con.prepareStatement("SELECT notify_then_sleep()");
+    final AtomicReference<SQLWarning> warning = new AtomicReference<SQLWarning>();
+    ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1);
+    ScheduledFuture<?> future = executorService.schedule(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          warning.set(preparedStatement.getWarnings());
+        } catch (SQLException e) {
+          fail("Exception thrown: " + e.getMessage());
```

Comment:
Yeah, no real reason not to if I use a `Callable`, but otherwise I cant throw a checked exception from a `Runnable`